net/golioth/rpc.c

```diff
@@ -0,0 +1,253 @@
+/*
+ * Copyright (c) 2022 Golioth, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <net/golioth/rpc.h>
+#include <logging/log.h>
+#include <qcbor/posix_error_map.h>
+#include <qcbor/qcbor.h>
+#include <qcbor/qcbor_spiffy_decode.h>
+#include <stdio.h>
+
+#include "coap_utils.h"
+
+/*
+ * Request:
+ * {
+ *      "id": "id_string,
+ *      "method": "method_name_string",
+ *      "params": [...]
+ * }
+ *
+ * Response:
+ * {
+ *      "id": "id_string",
+ *      "statusCode": integer,
+ *      "detail": {...}
+ * }
+ */
+
+LOG_MODULE_DECLARE(golioth);
+
+#define GOLIOTH_RPC_PATH ".rpc"
+#define GOLIOTH_RPC_STATUS_PATH ".rpc/status"
+#define GOLIOTH_RPC_MAX_RESPONSE_LEN 256
+
+struct golioth_rpc {
+	const char *method;
+	golioth_rpc_cb_fn callback;
+	void *callback_arg;
+};
+
+static struct golioth_client *_client;
+static struct coap_reply _observe_reply;
+static struct golioth_rpc _rpcs[CONFIG_GOLIOTH_RPC_MAX_NUM_METHODS];
+static int _num_rpcs;
```

Comment:
I think that those should be moved to `struct golioth_client`, as so far there was no restriction on using multiple `golioth_client` instances.

Additionally `_rpcs` and `_num_rpcs` should probably be protected with some mutex or something. Or if that is not the case, then maybe we should put restrictions on when (from which thread, which calllback) `golioth_rpc_register()` can be called, so that there are no race conditions between threads.

Reply:
I'll add mutex protection.